File: backend/src/main/java/com/cosmetics/ecommerce/service/OrderService.java
package com.cosmetics.ecommerce.service;

import org.springframework.data.domain.Page;

import com.cosmetics.ecommerce.dto.OrderDetailResponseDTO;
import com.cosmetics.ecommerce.dto.OrderListDTO;
import com.cosmetics.ecommerce.dto.OrderRequestDTO;
import com.cosmetics.ecommerce.dto.OrderResponseDTO;
import com.cosmetics.ecommerce.dto.OrderStatusUpdateResponseDTO;
import com.cosmetics.ecommerce.dto.UpdateOrderStatusRequestDTO;

/**
 * Interface định nghĩa các nghiệp vụ liên quan đến đơn hàng.
 *
 * Bao gồm:
 * - Đặt hàng (checkout)
 * - Xem danh sách đơn hàng của khách hàng
 * - Admin xem danh sách đơn hàng
 * - Admin xem chi tiết đơn hàng
 * - Admin cập nhật trạng thái đơn hàng
 *
 * Ý nghĩa:
 * - Tách biệt tầng Service và Controller
 * - Giúp dễ mở rộng và test (có thể mock khi viết unit test)
 */
public interface OrderService {
    /**
     * Thực hiện chức năng đặt hàng cho người dùng.
     *
     * @param userId  ID của người dùng
     * @param request Thông tin đặt hàng (địa chỉ, phương thức thanh toán)
     * @return Thông tin đơn hàng sau khi đặt thành công
     */
    OrderResponseDTO placeOrder(Integer userId, OrderRequestDTO request);

    /**
     * Lấy danh sách đơn hàng của người dùng.
     *
     * @param userId  ID người dùng
     * @param status  Trạng thái đơn hàng cần lọc, có thể null
     * @param keyword Từ khóa tìm kiếm, có thể null
     * @param page    Số trang cần lấy
     * @param size    Số lượng đơn hàng trên mỗi trang
     * @param sortBy  Trường dùng để sắp xếp
     * @param sortDir Chiều sắp xếp
     * @return Danh sách đơn hàng của người dùng theo dạng phân trang
     */
    Page<OrderResponseDTO> getMyOrders(
        Integer userId,
        String status,
        String keyword,
        int page,
        int size,
        String sortBy,
        String sortDir
    );

    /**
     * Lấy danh sách đơn hàng dành cho Admin.
     *
     * @param status        Trạng thái đơn hàng cần lọc, có thể null
     * @param keyword       Từ khóa tìm kiếm, có thể null
     * @param paymentMethod Phương thức thanh toán cần lọc, có thể null
     * @param paymentStatus Trạng thái thanh toán cần lọc, có thể null
     * @param page          Số trang cần lấy
     * @param size          Số lượng đơn hàng trên mỗi trang
     * @param sortBy        Trường dùng để sắp xếp
     * @param sortDir       Chiều sắp xếp
     * @return Danh sách đơn hàng dành cho Admin theo dạng phân trang
     */
    Page<OrderListDTO> getAdminOrders(
        String status, 
        String keyword,
        String paymentMethod,
        String paymentStatus, 
        int page, 
        int size,
        String sortBy,
        String sortDir
    );

    /**
     * Lấy thông tin chi tiết của một đơn hàng (dành cho Admin).
     *
     * @param orderId ID của đơn hàng
     * @return Thông tin chi tiết đơn hàng
     */
    OrderDetailResponseDTO getOrderDetailForAdmin(Integer orderId);

    /**
     * Cập nhật trạng thái của đơn hàng (dành cho Admin).
     *
     * @param orderId ID của đơn hàng
     * @param request Trạng thái mới cần cập nhật
     * @return Thông tin kết quả cập nhật trạng thái
     */
    OrderStatusUpdateResponseDTO updateOrderStatus(Integer orderId, UpdateOrderStatusRequestDTO request);

    /**
     * Lấy chi tiết đơn hàng của người dùng.
     *
     * @param userId  ID người dùng
     * @param orderId ID đơn hàng
     * @return Thông tin chi tiết đơn hàng
     */
    OrderDetailResponseDTO getOrderDetailForCustomer(Integer userId, Integer orderId);

    /**
     * Xác nhận thanh toán thành công cho đơn hàng COD.
     *
     * @param orderId ID đơn hàng COD cần xác nhận thanh toán
     */
    void confirmCodPayment(Integer orderId);
}

File: frontend/src/features/orders/orderService.js
import { orderApi } from "./orderApi";

/**
 * orderService chứa các hàm nghiệp vụ phía frontend liên quan đến đơn hàng.
 *
 * File này gọi orderApi để gửi request đến backend,
 * sau đó lấy response.data trả về cho component sử dụng.
 *
 * Khác với orderApi:
 * - orderApi: chỉ gọi HTTP request
 * - orderService: lấy dữ liệu cần dùng từ response
 */
export const orderService = {
    /**
     * Gửi yêu cầu đặt hàng.
     *
     * @param data Dữ liệu đặt hàng
     * @returns Dữ liệu đơn hàng vừa được tạo
     */
    async placeOrder(data) {
        const response = await orderApi.placeOrder(data); // Gọi API đặt hàng.
        return response.data; // Trả về phần data từ response để component dùng trực tiếp.
    },

    /**
     * Lấy danh sách đơn hàng của khách hàng đang đăng nhập.
     *
     * @param params Tham số tìm kiếm, lọc, sắp xếp và phân trang
     *
     * @returns Danh sách đơn hàng của khách hàng
     */
    async getMyOrders(params) {
        const response = await orderApi.getMyOrders(params); // Gọi API lấy danh sách đơn hàng của khách.
        return response.data; // Trả về dữ liệu backend gửi về.
    },

    /**
     * Lấy chi tiết một đơn hàng của khách hàng đang đăng nhập.
     *
     * @param orderId ID của đơn hàng cần xem chi tiết
     *
     * @returns Chi tiết đơn hàng
     */
    async getMyOrderDetail(orderId) {
        const response = await orderApi.getMyOrderDetail(orderId); // Gọi API lấy chi tiết đơn hàng phía khách hàng.
        return response.data; // Trả về dữ liệu chi tiết đơn hàng.
    },

    /**
     * Lấy danh sách đơn hàng phía Admin.
     *
     * @param params Tham số tìm kiếm, lọc, sắp xếp và phân trang
     * @returns Danh sách đơn hàng cho Admin
     */
    async getAdminOrders(params) {
        const response = await orderApi.getAdminOrders(params); // Gọi API lấy danh sách đơn hàng phía Admin.
        return response.data; // Trả về dữ liệu danh sách đơn hàng.
    },

    /**
     * Lấy chi tiết một đơn hàng phía Admin.
     *
     * @param orderId ID của đơn hàng cần xem chi tiết
     * @returns Chi tiết đơn hàng cho Admin
     */
    async getAdminOrderDetail(orderId) {
        const response = await orderApi.getAdminOrderDetail(orderId); // Gọi API lấy chi tiết đơn hàng phía Admin.
        return response.data; // Trả về dữ liệu chi tiết đơn hàng.
    },

    /**
     * Cập nhật trạng thái đơn hàng phía Admin.
     *
     * @param orderId ID của đơn hàng cần cập nhật
     * @param status Trạng thái mới của đơn hàng
     * @returns Kết quả cập nhật trạng thái
     */
    async updateAdminOrderStatus(orderId, status) {
        const response = await orderApi.updateAdminOrderStatus(orderId, status);
        return response.data;
    },

    /**
     * Xác nhận thanh toán thành công cho đơn COD.
     *
     * @param orderId ID của đơn hàng COD cần xác nhận thanh toán
     * @returns Kết quả xác nhận thanh toán COD
     */
    async confirmCodPayment(orderId) {
        const response = await orderApi.confirmCodPayment(orderId);
        return response.data;
    }
};